File: python/day01.py
import logging

from advent import Solver

log = logging.getLogger("day.01")


# https://adventofcode.com/2022/day/1
# quite simple, of course just go through it in one pass of the file.
# both solutions can be calculated while reading the file
class Solution(Solver):
    def __init__(self):
        self.elf = 0
        self.elf_calories = 0
        self.calories = []

    def parse(self, line: str):
        if not line:
            # print(f"Elf n. {self.elf} holding {self.elf_calories} calories")
            self.calories.append([self.elf, self.elf_calories])
            self.elf_calories = 0
            self.elf += 1
            return
        self.elf_calories += int(line)

    def solve(self):
        self.calories.append([self.elf, self.elf_calories])
        self.calories.sort(key=lambda x: x[1], reverse=True)
        part1 = self.calories[0][1]
        log.info(f"[1] Saw {self.elf} elves: maximum: {self.calories[0]}")
        top3 = [x[1] for x in self.calories[0:3]]
        part2 = sum(top3)
        log.info(f"[2] First 3 elves: {part2}")
        return str(part1), str(part2)

File: rust/src/bin/aoc2022/solutions/day01.rs
// What did I learn?
// Vec basic APIs and a bit of functional/iterator constructs.

use std::str::FromStr;
use log::info;
use crate::solutions::Solver;

struct Elf {
    calories: u32,
}

pub struct Solution {
    elf: u32,
    elf_calories: u32,
    calories: Vec<Elf>,
}

impl Solution {
    pub(crate) fn new() -> Solution {
        Solution {
            elf: 0,
            elf_calories: 0,
            calories: Vec::new(),
        }
    }
}

impl Solver for Solution {
    fn parse(&mut self, line: &str) {
        if line.is_empty() {
            self.calories.push(Elf{calories: self.elf_calories});
            self.elf += 1;
            self.elf_calories = 0;
        } else {
            self.elf_calories += u32::from_str(line).unwrap();
        }
    }

    fn solve(&mut self) -> Option<(String, String)> {
        info!("solve day01");
        self.calories.push(Elf{calories: self.elf_calories});
        self.calories.sort_by_key(|elf| elf.calories);
        self.calories.reverse();
        info!("[1] Saw {} elves: maximum: {}", self.calories.len(), self.calories[0].calories);
        let top3: u32 = self.calories.iter().take(3).map(|elf| elf.calories).sum();
        info!("[2] First 3 elves: {top3}");
        Some((self.calories[0].calories.to_string(), top3.to_string()))
    }
}